myhpi/core/context.py

```diff
@@ -1,9 +1,65 @@
-from wagtail.core.models import Page, Site
+from django.contrib.auth.models import Group
+from django.db.models import Q
+from wagtail.core.models import Site
+
+from .models import BasePage
 
 
 def base_context(request):
+
+    # How wagtail page trees work: https://www.accordbox.com/blog/how-to-create-and-manage-menus-in-wagtail/
+
+    # Fetch all pages
     root_page = getattr(Site.find_for_request(request), "root_page", None)
+    pages = BasePage.objects.in_menu().live()
+    pages_visible_for_user = []
+    path_map = {}
+    depth_levels = set()
+
+    if not root_page:
+        return {"root_page": None, "all_pages": None}
+
+    # Determine all pages the user may view based on his groups
+    usergroups = {}
+    if getattr(request.user, "_ip_range_group_name", False):
+        # join user groups together with the groups they have based on their IP address
+        usergroups = Group.objects.filter(
+            Q(name=request.user._ip_range_group_name) | Q(id__in=request.user.groups.all())
+        )
+    else:
+        # use the users groups only
+        usergroups = request.user.groups.all()
+    pages_visible_for_user = pages.filter(visible_for__in=usergroups)
+
+    # Add additional properties to each page and save in path_map
+    # The path of each page is represented by a string, each level labeled with 4 chars
+    for page in pages:
+        page.menu_children = []
+        page.is_leaf = True
+        path_map[page.path] = page
+        depth_levels.add(page.depth)
+
+    # Iterate over all depth of the tree, going from leaves to root
```

Comment:
```suggestion
    # Iterate over entire depth of the tree, going from leaves to root
```
Or what is this comment supposed to say?

Reply:
Also why go from leaves to root and not the other way?

Reply:
because when none of a page's child pages are visible, the page itself is also not supposed to be visible.
going from root to leaves would require backtracking which I wanted to avoid here.